builder/chroots.go

```diff
@@ -80,6 +82,416 @@ func readBuildChrootsConfig(path string) (*buildChrootsConfig, error) {
+	return re.ReplaceAllLiteralString(path, "/usr/lib64/")
+}
+
+func addFileAndPath(destination map[string]bool, absPathsToFiles ...string) {
+	for _, file := range absPathsToFiles {
+		if isBannedPath(file) {
+			continue
+		}
+
+		path := "/"
+		dir := filepath.Dir(file)
+		// skip first empty string so we don't add '/'
+		for _, part := range strings.Split(dir, "/")[1:] {
+			path = filepath.Join(path, part)
+			destination[path] = true
+		}
+		destination[file] = true
+	}
+}
+
+func addOsCoreSpecialFiles(bundle *bundle) {
+	filesToAdd := []string{
+		"/usr/lib/os-release",
+		"/usr/share/clear/version",
+		"/usr/share/clear/versionstamp",
+	}
+
+	addFileAndPath(bundle.Files, filesToAdd...)
+}
+
+func addUpdateBundleSpecialFiles(b *Builder, bundle *bundle) {
+	filesToAdd := []string{
+		"/usr/share/defaults/swupd/contenturl",
+		"/usr/share/defaults/swupd/versionurl",
+		"/usr/share/defaults/swupd/format",
+	}
+
+	if _, err := os.Stat(b.Cert); err == nil {
+		filesToAdd = append(filesToAdd, "/usr/share/clear/update-ca/Swupd_Root.pem")
+	}
+
+	addFileAndPath(bundle.Files, filesToAdd...)
+}
+
+func resolveFilesForBundle(bundle *bundle, packagerCmd []string) error {
+	bundle.Files = make(map[string]bool)
+	for p := range bundle.DirectPackages {
+		queryString := merge(packagerCmd, "repoquery", "-l", "--quiet", p)
+		outBuf, err := helpers.RunCommandOutput(false, queryString[0], queryString[1:]...)
+		if err != nil {
+			return err
+		}
+		for _, f := range strings.Split(outBuf.String(), "\n") {
+			if len(f) > 0 {
+				addFileAndPath(bundle.Files, resolveFileName(f))
+			}
+		}
+	}
+	addFileAndPath(bundle.Files, fmt.Sprintf("/usr/share/clear/bundles/%s", bundle.Name))
+	fmt.Printf("Bundle %s\t%d files\n", bundle.Name, len(bundle.Files))
+	return nil
+}
+
+func resolveFiles(numWorkers int, set bundleSet, packagerCmd []string) error {
+	var err error
+	var wg sync.WaitGroup
+	fmt.Printf("Resolving files using %d workers\n", numWorkers)
+	wg.Add(numWorkers)
+	bundleCh := make(chan *bundle)
+	errorCh := make(chan error, numWorkers)
+	fileWorker := func() {
+		for bundle := range bundleCh {
+			fmt.Printf("processing %s\n", bundle.Name)
+			err = resolveFilesForBundle(bundle, packagerCmd)
+			if err != nil {
+				errorCh <- err
+			}
+		}
+		wg.Done()
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go fileWorker()
+	}
+
+	for _, bundle := range set {
+		select {
+		case bundleCh <- bundle:
+		case err = <-errorCh:
+			// Break as soon as there is a failure.
+			break
+		}
+	}
+	close(bundleCh)
+	wg.Wait()
+
+	// Sending loop might finish before any goroutine could send an error back, so check for
+	// error again after they are all done.
+	if err == nil && len(errorCh) > 0 {
+		err = <-errorCh
+	}
+	return err
+}
+
+func parseNoopInstall(installOut string) []string {
+	pkgDeps := []string{}
+	inBlock := false
+	re := regexp.MustCompile("[[:space:]]([[:graph:]]*)[[:space:]]")
+	for _, line := range strings.Split(installOut, "\n") {
```

Comment:
Why not split the installOut string into 3 parts first, at the "Installing dependencies" and the first blank line, and then process the lines in the middle part?

Using a regular expression to split the line apart is a bit heavyweight!

If you want to keep using a regular expression, compile it only (at most) once per run of mixer, not once per call to parseNoopInstall.

Reply:
done.